spark3/src/main/java/org/apache/iceberg/actions/Spark3CreateAction.java

```diff
@@ -0,0 +1,310 @@
+      throw e;
+    }
+
+    if (sessionCatalogReplacement) {
+      Identifier replacementTable = Identifier.of(destTableName.namespace(), destTableName.name() + REPLACEMENT_NAME);
+      try {
+        stagingCatalog.dropTable(destTableName);
+        stagingCatalog.renameTable(replacementTable, destTableName);
+      } catch (NoSuchTableException e) {
+        LOG.error("Cannot migrate, replacement table is missing. Attempting to recreate source table", e);
+        try {
+          stagingCatalog.createTable(sourceTableName, sourceTable.schema(),
+              Spark3Util.toTransforms(sourcePartitionSpec), JavaConverters.mapAsJavaMap(sourceTable.properties()));
+        } catch (TableAlreadyExistsException tableAlreadyExistsException) {
+          Log.error("Cannot recreate source table. Source table has already been recreated", e);
+          throw new RuntimeException(e);
+        } catch (NoSuchNamespaceException noSuchNamespaceException) {
+          Log.error("Cannot recreate source table. Source namespace has been removed, cannot recreate", e);
+          throw new RuntimeException(e);
+        }
+      } catch (TableAlreadyExistsException e) {
+        Log.error("Cannot migrate, Source table was recreated before replacement could be moved. " +
+            "Attempting to remove replacement table.", e);
+        stagingCatalog.dropTable(replacementTable);
+        stagedTable.abortStagedChanges();
+      }
+    }
+
+    return numMigratedFiles;
+  }
+
+  @Override
+  public CreateAction set(Map<String, String> properties) {
+    this.additionalProperties.putAll(properties);
+    return this;
+  }
+
+  @Override
+  public CreateAction set(String key, String value) {
+    this.additionalProperties.put(key, value);
+    return this;
+  }
+
+  private static void validateSourceTable(CatalogTable sourceTable) {
+    String sourceTableProvider = sourceTable.provider().get().toLowerCase(Locale.ROOT);
+
+    if (!ALLOWED_SOURCES.contains(sourceTableProvider)) {
+      throw new IllegalArgumentException(
+          String.format("Cannot create an Iceberg table from source provider: %s", sourceTableProvider));
+    }
+    if (sourceTable.storage().locationUri().isEmpty()) {
+      throw new IllegalArgumentException("Cannot create an Iceberg table from a source without an explicit location");
+    }
+  }
+
+  private static Map<String, String> extraIcebergTableProps(String tableLocation, String metadataLocation) {
+    if (tableLocation != null && metadataLocation != null) {
+      return ImmutableMap.of(
+          TableProperties.WRITE_METADATA_LOCATION, metadataLocation,
+          TableProperties.WRITE_NEW_DATA_LOCATION, tableLocation,
+          "migrated", "true"
+      );
+    } else {
+      return ImmutableMap.of("migrated", "true");
+    }
+  }
+
+  private static StagingTableCatalog checkDestinationCatalog(CatalogPlugin catalog) {
+    if (!(catalog instanceof SparkSessionCatalog) && !(catalog instanceof SparkCatalog)) {
```

Comment:
Is it enough to check if it is instance of `BaseCatalog` we introduced recently?

Reply:
Yep, Sorry this PR has been in progress for a very long time, I haven't kept up with all the other changes.

Reply:
We can't actually do this with the current package structure since baseCatalog is package private in Spark and this is in Actions. We could always move spark's action package though